test/files/run/t10788.scala

```diff
@@ -0,0 +1,16 @@
+object Test {
+  class Box[T](t: T) {
+    def foo: T = {
+      println("effect")
+      t
+    }
+  }
+
+  object Box {
+    def apply(x: String): Box[x.type] = new Box[x.type](x)
+  }
+
+  def main(args: Array[String]): Unit = {
+    val bar = Box("foo").foo
+  }
+}
```

Comment:
Here's an even simpler test variation that is also fixed by this change.

```scala
class C {
  println("effect")
  final val const = 42
}

object Test {
  def main(args: Array[String]) {
    val x = new C().const
  }
}
```

Reply:
But I wonder if rather than fully deconsting we should instead convert `ConstantType`-s to `LiteralType`-s.

Reply:
That is what `deconst` on a `FoldableConstantType` does.